### ISSUE TITLE
Client-side certificate validations issues

### ISSUE THREAD

**Issue by asendra** [ckrey/MQTT-Client-Framework#96]
I´m having trouble getting to work client side validation using self-signed certificates.
We have our own CA Authority which generates a server certificate which I can validate via MQTTSSLSecurityPolicy methods.
The same CA also generates the client certificate. It's a pem certificate and private key which I combine in a PKCS12 file and import it using the built in function of the library and assign the array to the MQTTSession.certificates property.

As far as I can see, that is the proper way to do it but I get a "CFNetwork SSLHandshake failed (-9824 -> -9829)" error. 
Also, debugging the connection from the server side using openssl gives the error "peer did not return a certificate".

If I remove the server check for a client certificate, the connection works.

Does the client certificate needs to be generate in a specific way? Using the pem generated client certificates from a python script establishes connection perfectly so in theory the certificates are correctly generated.

Any ideas?

Code:
```objectivec
NSString *serverCertPath = [[NSBundle mainBundle] pathForResource:@"server" ofType:@"der"];
NSData *serverCertData = [[NSData alloc] initWithContentsOfFile:serverCertPath];
        
MQTTSSLSecurityPolicy *policy = [MQTTSSLSecurityPolicy policyWithPinningMode:MQTTSSLPinningModeCertificate];
policy.validatesDomainName = YES;
policy.allowInvalidCertificates = YES;
policy.pinnedCertificates = @[serverCertData];

[_mqttManager connectTo:@"hostname.com"
                           port:8883
                            tls:YES
                      keepalive:3
                          clean:YES
                           auth:YES
                           user:@"[some user]"
                           pass:@"[some password]"
                           will:YES
                      willTopic:@"[some topic]"
                        willMsg:[@"[some message]" dataUsingEncoding:NSUTF8StringEncoding]
                        willQos:MQTTQosLevelAtMostOnce
                 willRetainFlag:NO
                   withClientId:@"[some client id]"
                 securityPolicy:policy
                   certificates:[MQTTCFSocketTransport clientCertsFromP12:[[NSBundle mainBundle] pathForResource:@"mqtt-client-cert" ofType:@"p12"]
                                                               passphrase:@"mqtt"]];
```

**Comment by asendra** [comment]
So, I ended up figuring out what was going on. Should have been easy to spot the bug, but for some reason, I didn't check it.

Using the above mentioned `connectTo:port:....securityPolicy:certificates:` method to configure and start the connection, assigns the parameters to member properties of MQTTSessionManager, and initialises a MQTTSession passing the same parameters. MQTTSession, in it's initialiser, also assigns the variables to it's own member properties. 

The bug is that the `setCertificates:` method checks first if exists already a transport initialised, and only stores the certificates array if it does, which at this point in the code, transport hasn't been initialised yet.

Transport isn't initialised until the call to method `connectToHost:port:usingSSL:connectHandler:` is made from `connectToInternal`

Basically, the certificates property of MQTTSession is a stub for accessing the certificates property in the session's transport, but if this doesn't exist yet (uninitialised), then the certificates array passed from the manager is lost, and when the MQTTSession instance initialises the transport and tries to pass its certificates, its empty.

the correct way to do it would be to either initialise the transport at the same time as the MQTTSession, or hold the certificates property in the MQTTSession instance.

**Comment by ckrey** [comment]
@asendra Thanks for finding the inconsistency. Will you send me a pull request, or do you want to wait until I have time to fix it?

**Comment by ckrey** [comment]
Thank you @asendra for your post. It took me a while to fully understand what is actually wrong. I will fix immediately.

It took me a while because the problem happens only with the legacy initializer methods.

Problem shows up here too: owntracks/ios#379

**Comment by ckrey** [comment]
Fixed

**Comment by Reetalibrarian** [comment]
I got similar problem. I keep receiving errors like "OpenSSL Error: error:140940E5:SSL routines:SSL3_READ_BYTES:ssl handshake failure"

We previously can connect to the server if server if open for anyone without checking cert/key. 

code:
`
self.mqttSession = [[MQTTSession alloc] init];
MQTTSSLSecurityPolicy* securityPolicy = [MQTTSSLSecurityPolicy policyWithPinningMode:MQTTSSLPinningModeCertificate];
NSData *rootCA = [NSData dataWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"root_ca" ofType:@"der"]];
securityPolicy.pinnedCertificates = @[rootCA];
securityPolicy.allowInvalidCertificates = YES;
securityPolicy.validatesCertificateChain = NO;
securityPolicy.validatesDomainName = NO;

self.mqttSession.certificates = [MQTTCFSocketTransport clientCertsFromP12:[[NSBundle mainBundle] pathForResource:@"client_certificate" ofType:@"p12"] passphrase:@""];
self.mqttSession.userName = @"name";
self.mqttSession.password = nil;
self.mqttSession.keepAliveInterval = 60;
self.mqttSession.cleanSessionFlag = YES;
self.mqttSession.protocolLevel = 4;
self.mqttSession.delegate = self;

[self.mqttSession connectToHost:SERVICE_HOST port:8883 usingSSL:YES];
`

Log from server
OpenSSL Error: error:140940E5:SSL routines:SSL3_READ_BYTES:ssl handshake failure
(error code : 9807)

if i change the code to be 
`
self.mqttSession = [[MQTTSession alloc] init];
MQTTSSLSecurityPolicy* securityPolicy = [MQTTSSLSecurityPolicy policyWithPinningMode:MQTTSSLPinningModeCertificate];
    NSData *rootCA = [NSData dataWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"roote_ca" ofType:@"der"]];
    securityPolicy.pinnedCertificates = @[rootCA];

    NSString *certificate = [[NSBundle mainBundle] pathForResource:@"candyhouse_root_ca" ofType:@"der"];
    self.mqttSession.securityPolicy = [MQTTSSLSecurityPolicy policyWithPinningMode:MQTTSSLPinningModeCertificate];
    self.mqttSession.securityPolicy.pinnedCertificates = @[[NSData dataWithContentsOfFile:certificate]];
    NSString *path = [[NSBundle mainBundle] pathForResource:@"client_cert" ofType:@"p12"];
    self.mqttSession.certificates = [MQTTCFSocketTransport clientCertsFromP12:path passphrase:@""];
    [self.mqttSession connectToHost:SERVICE_HOST port:8883 usingSSL:YES];
`

Server log will be:
OpenSSL Error: error:140890B2:SSL routines:SSL3_GET_CLIENT_CERTIFICATE:no certificate returned
(error code : 9831)

These logs confused me, and I'm not sure if I miss something.

**Comment by Reetalibrarian** [comment]
I got similar problem. I keep receiving errors like "OpenSSL Error: error:140940E5:SSL routines:SSL3_READ_BYTES:ssl handshake failure"

We previously can connect to the server if server is open for anyone without checking cert/key.

code:

```objective-c
self.mqttSession = [[MQTTSession alloc] init]; 

// Security Policy
MQTTSSLSecurityPolicy* securityPolicy = [MQTTSSLSecurityPolicy policyWithPinningMode:MQTTSSLPinningModeCertificate]; 
NSData *rootCA = [NSData dataWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"root_ca" ofType:@"der"]]; securityPolicy.pinnedCertificates = @[rootCA]; 
securityPolicy.allowInvalidCertificates = YES; 
securityPolicy.validatesCertificateChain = NO; 
securityPolicy.validatesDomainName = NO;

// Security Policy Transport
MQTTSSLSecurityPolicyTransport *sslSecurityPolicyTransport = [[MQTTSSLSecurityPolicyTransport alloc] init];
sslSecurityPolicyTransport.tls = YES;
sslSecurityPolicyTransport.securityPolicy = securityPolicy;

self.mqttSession.transport = sslSecurityPolicyTransport;
self.mqttSession.certificates = [MQTTCFSocketTransport clientCertsFromP12:[[NSBundle mainBundle] pathForResource:@"client_certificate" ofType:@"p12"] passphrase:@""]; 
self.mqttSession.userName = @"name"; self.mqttSession.password = nil; 
self.mqttSession.keepAliveInterval = 60; 
self.mqttSession.cleanSessionFlag = YES; 
self.mqttSession.protocolLevel = 4; 
self.mqttSession.delegate = self; 
[self.mqttSession connectToHost:SERVICE_HOST port:8883 usingSSL:YES];
```
Log from server
##### OpenSSL Error: error:140940E5:SSL routines:SSL3_READ_BYTES:ssl handshake failure(error code : 9807)

------------------------------------------------------------------------------------

if i change the code to be 

```objective-c
self.mqttSession = [[MQTTSession alloc] init]; 

// Server-Certificate
NSString *certificate = [[NSBundle mainBundle] pathForResource:@"root_ca" ofType:@"der"];
self.mqttSession.securityPolicy = [MQTTSSLSecurityPolicy policyWithPinningMode:MQTTSSLPinningModeCertificate];
self.mqttSession.securityPolicy.pinnedCertificates = @[[NSData dataWithContentsOfFile:certificate]];

// Client-Certificate
NSString *path = [[NSBundle mainBundle] pathForResource:@"Client_certificate" ofType:@"p12"];
self.mqttSession.certificates = [MQTTCFSocketTransport clientCertsFromP12:path passphrase:@""];
[self.mqttSession connectToHost:SERVICE_HOST port:8883 usingSSL:YES];
 ```

Server log will be:
##### OpenSSL Error: error:140890B2:SSL routines:SSL3_GET_CLIENT_CERTIFICATE:no certificate returned (error code : 9831)

These logs confused me, and I'm not sure if I miss something.

**Comment by aratalcordtech** [comment]
Hi, I can't seem to use the certificates correctly in the latest version. Could somebody please help me out.
https://github.com/novastone-media/MQTT-Client-Framework/issues/545
